### PR TITLE
L2 bridge mantle

### PR DESCRIPTION
**Pull request by sergeyWh1te**
## How does it work.

This PR introduces a new bot operating on the Mantle network.
The existing bot functions on two networks: ETH mainnet and Mantle.
Here's how it operates: The bot monitors new blocks on the ETH mainnet.
Since Forta doesn't currently support Mantle, the bot reads blocks on the L2 network (Mantle) and stores the latest one
in an in-memory cache.
When the bot reads the next block on ETH, it also retrieves a segment of Mantle blocks (cachedBlock, LatestBlock) from
the cache.

## Supported chains

- Ethereum mainnet, Mantle network

## Alerts

1. Bridge events
   1. 🚨 Mantle L2 Bridge: Role Admin changed
   2. ⚠️ Mantle L2 Bridge: Role granted
   3. ⚠️ Mantle L2 Bridge: Role revoked
   4. ✅ Mantle L2 Bridge: Deposits Enabled
   5. ❌ Mantle L2 Bridge: Deposits Disabled
   6. ✅ Mantle L2 Bridge: Withdrawals Enabled
   7. ❌ Mantle L2 Bridge: Withdrawals Disabled
   8. 🚨 Mantle L2 Bridge: Implementation initialized
2. Gov Events
   1. 🚨 Mantle Gov Bridge: Ethereum Governance Executor Updated
   2. 🚨 Mantle Gov Bridge: Guardian Updated
   3. ⚠️ Mantle Gov Bridge: Delay Updated
   4. ⚠️ Mantle Gov Bridge: Grace Period Updated
   5. ⚠️ Mantle Gov Bridge: Min Delay Updated
   6. ⚠️ Mantle Gov Bridge: Max Delay Updated
   7. ℹ Mantle Gov Bridge: Action set queued
   8. ℹ Mantle Gov Bridge: Action set executed
   9. ℹ Mantle Gov Bridge: Action set canceled
3. Proxy events
   1. 🚨 Mantle: Proxy ossified
   2. 🚨 Mantle: Proxy admin changed
   3. 🚨 Mantle: Proxy upgraded
   4. 🚨 Mantle: Proxy beacon upgraded
4. Monitor Withdrawals
   1. ⚠️ Mantle: Huge withdrawals during the last ...
5. Agent balance
   1. Added wstETH: 🚨🚨🚨 Mantle bridge balance mismatch 🚨🚨🚨

## Proofs of addreses and constants got from https://docs.lido.fi/deployed-contracts/:
      
1. L1ERC20TokenBridge Impl : [0x6fBBe1Af52D22557D7F161Dc5952E306F4742e23](https://etherscan.io/address/0x2D001d79E5aF5F65a939781FE228B267a8Ed468B)
2. L1ERC20TokenBridge Proxy: [0x2D001d79E5aF5F65a939781FE228B267a8Ed468B](https://etherscan.io/address/0x6fBBe1Af52D22557D7F161Dc5952E306F4742e23)
 3. ERC20Bridged Impl: [0x1FaBaAec88198291A4efCc85Cabb33a3785165ba](https://explorer.mantle.xyz/address/0x1FaBaAec88198291A4efCc85Cabb33a3785165ba)
 4. ERC20Bridged Proxy: [0x458ed78EB972a369799fb278c0243b25e5242A83](https://explorer.mantle.xyz/address/0x458ed78EB972a369799fb278c0243b25e5242A83)
 5. L2ERC20TokenBridge Impl: [0xf10A7ffC613a9b23Abc36167925A375bf5986181](https://explorer.mantle.xyz/address/0xf10A7ffC613a9b23Abc36167925A375bf5986181)
6. L2ERC20TokenBridge Proxy: [0x9c46560D6209743968cC24150893631A39AfDe4d](https://explorer.mantle.xyz/address/0x9c46560D6209743968cC24150893631A39AfDe4d)
 7. OptimismBridgeExecutor: [0x3a7b055bf88cdc59d20d0245809c6e6b3c5819dd](https://explorer.mantle.xyz/address/0x3a7b055bf88cdc59d20d0245809c6e6b3c5819dd)

### CODE REVIEW

File: l2-bridge-mantle/README.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+   1. 🚨 Mantle: Proxy ossified
+   2. 🚨 Mantle: Proxy admin changed
+   3. 🚨 Mantle: Proxy upgraded
+   4. 🚨 Mantle: Proxy beacon upgraded
+4. Monitor Withdrawals
+   1. ⚠️ Mantle: Huge withdrawals during the last ...
+5. Agent balance
+   1. Added wstETH: 🚨🚨🚨 Mantle bridge balance mismatch 🚨🚨🚨
+
+## How to setup new bot and network
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply l2-bridge-mantle/README.md:48]: would suggest having this guide somewhere in the top level file and have the embedded link on it within each L2 bot

File: l2-bridge-mantle/src/config/bot-config.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "Mantle": {
+    "RpcUrlBase64": "aHR0cHM6Ly9tYW50bGUucHVibGljbm9kZS5jb20="
```
Comment (TheDZhon) [inline l2-bridge-mantle/src/config/bot-config.json:3]: I am a bit confused

File: l2-bridge-mantle/src/generated/factories/L2ERC20TokenBridge__factory.ts
Comment (TheDZhon) [inline l2-bridge-mantle/src/generated/factories/L2ERC20TokenBridge__factory.ts:1]: Do we really need to commit generated files?

File: l2-bridge-mantle/src/utils/constants.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+export const MANTLE_WST_ETH_BRIDGED: MANTLE_WST_ETH_BRIDGED_TYPE = {
+  name: 'MANTLE_WST_ETH_BRIDGED_ADDRESS',
+  address: MANTLE_WST_ETH_BRIDGED_ADDRESS,
+}
+
+export const DEPOSITS_DISABLER_ROLE: Role = {
+  name: 'DEPOSITS DISABLER ROLE',
+  address: '0x63f736f21cb2943826cd50b191eb054ebbea670e4e962d0527611f830cd399d6',
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply l2-bridge-mantle/src/utils/constants.ts:33]: applicable here and there

File: template-v2/src/abi/L2Bridge.json
Comment (TheDZhon) [inline template-v2/src/abi/L2Bridge.json:1]: won't be a universal interface tho

File: template-v2/src/config/bot-config.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "Base": {
+    "RpcUrlBase64": "aHR0cHM6Ly9iYXNlLW1haW5uZXQuZy5hbGNoZW15LmNvbS92Mi9sQ0RseWdMNk00eHpVdF9KTEQxMFo3MW9pMkRfeVpiNA"
```
Comment (TheDZhon) [inline template-v2/src/config/bot-config.json:3]: alchemy key :eyes:

File: template-v2/src/generated/index.ts
Comment (TheDZhon) [inline template-v2/src/generated/index.ts:1]: Are we totally sure that the `generated` contents should be committed at all I mean?

**Review by TheDZhon — CHANGES_REQUESTED** [review]
Thank you for bringing the refined architecture vision. :guitar: 

I would suggest addressing the findings highlighted before going further :pray:

File: l2-bridge-balance/src/agent-balance.ts [hunk 5]
```diff
@@ -48,8 +47,14 @@ export async function handleBlock(blockEvent: BlockEvent) {
...
       BRIDGE_PARAMS_WSTETH.ZkSync,
     ),
+    handleBridgeBalanceWstETH(
+      blockEvent,
+      findings,
+      BRIDGE_PARAMS_WSTETH.Mantle,
+    ),
     handleBridgeBalanceLDO(blockEvent, findings, BRIDGE_PARAMS_LDO.Arbitrum),
     handleBridgeBalanceLDO(blockEvent, findings, BRIDGE_PARAMS_LDO.Optimism),
+    handleBridgeBalanceLDO(blockEvent, findings, BRIDGE_PARAMS_LDO.Optimism),
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-balance/src/agent-balance.ts:57]: fixed

File: l2-bridge-balance/src/constants.ts [hunk 6]
```diff
@@ -49,6 +50,12 @@ export const BRIDGE_PARAMS_WSTETH: BridgeParamsWstETH = {
     wstEthBridged: "0x703b52F2b28fEbcB60E1372858AF5b18849FE867",
     rpcUrl: config.ZkSync.RpcUrl,
   },
+  Mantle: {
+    name: "Mantle",
+    l1Gateway: "0x2D001d79E5aF5F65a939781FE228B267a8Ed468B",
+    wstEthBridged: "0x458ed78EB972a369799fb278c0243b25e5242A83",
+    rpcUrl: config.ZkSync.RpcUrl,
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-balance/src/constants.ts:57]: fixed

File: l2-bridge-mantle/src/generated/factories/L2ERC20TokenBridge__factory.ts
Comment (sergeyWh1te) [reply l2-bridge-mantle/src/generated/factories/L2ERC20TokenBridge__factory.ts:1]: Yes, we need

File: l2-bridge-mantle/src/utils/constants.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-mantle/src/utils/constants.ts:33]: fixed

File: template-v2/src/config/bot-config.json (hunk 4 above)
Comment (sergeyWh1te) [reply template-v2/src/config/bot-config.json:3]: fix

File: template-v2/src/generated/index.ts
Comment (sergeyWh1te) [reply template-v2/src/generated/index.ts:1]: yes

File: l2-bridge-mantle/src/config/bot-config.json (hunk 2 above)
Comment (sergeyWh1te) [reply l2-bridge-mantle/src/config/bot-config.json]: fixed

File: README.md [hunk 7]
```diff
@@ -32,25 +40,143 @@
 
 [reverted-tx-watcher](./reverted-tx-watcher) - Detection bot that monitors critical storage changes
 
-[template](./template) - Template for bot creation
+[template](./template) - Template for bot creation. (Forta version)
+
+[template-v2](./template-v2) - Template for bot creation. (Sergey Wh1te's version)
+
+## How to setup new bot and network on example from template-v2 folder or l2-bridge-mantle
```
Comment (TheDZhon) [inline README.md:47]: I am not totally convinced that we should populate top-level README with L2-specifics :warning:

File: l2-bridge-ethereum/src/constants.ts [hunk 8]
```diff
@@ -53,10 +53,17 @@ export const OPTIMISM_L1ERC20_TOKEN_BRIDGE =
   "0x76943c0d61395d8f2edf9060e1533529cae05de6";
 export const BASE_L1ERC20_TOKEN_BRIDGE =
   "0x9de443AdC5A411E83F1878Ef24C3F52C61571e72";
-export const ZKSYNC_L1ERC20_BRIDGE =
+export const ZKSYNC_L1ERC20_TOEKEN_BRIDGE =
```
Comment (TheDZhon) [inline l2-bridge-ethereum/src/constants.ts:56]:
```suggestion
export const ZKSYNC_L1ERC20_TOKEN_BRIDGE =
```

File: l2-bridge-ethereum/src/constants.ts [hunk 9]
```diff
@@ -75,7 +82,11 @@ export const L1_ERC20_TOKEN_GATEWAYS = [
   },
   {
     name: "ZkSync",
-    address: ZKSYNC_L1ERC20_BRIDGE,
+    address: ZKSYNC_L1ERC20_TOEKEN_BRIDGE,
```
Comment (TheDZhon) [inline l2-bridge-ethereum/src/constants.ts:85]:
```suggestion
    address: ZKSYNC_L1ERC20_TOKEN_BRIDGE,
```

File: README.md (hunk 7 above)
Comment (sergeyWh1te) [reply README.md:47]: moved

File: l2-bridge-ethereum/src/constants.ts (hunk 8 above)
Comment (sergeyWh1te) [reply l2-bridge-ethereum/src/constants.ts:56]: fix

File: l2-bridge-ethereum/src/constants.ts (hunk 9 above)
Comment (sergeyWh1te) [reply l2-bridge-ethereum/src/constants.ts:85]: fix

File: l2-bridge-ethereum/src/constants.ts [hunk 10]
```diff
@@ -126,7 +137,7 @@ export const LIDO_PROXY_CONTRACTS: LidoProxy[] = [
   },
   {
     name: "L1ERC20TokenBridge to ZkSync",
-    address: ZKSYNC_L1ERC20_BRIDGE,
+    address: ZKSYNC_L1ERC20_TOEKEN_BRIDGE,
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-ethereum/src/constants.ts:140]: fix

File: l2-bridge-mantle/src/utils/constants.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import { Role, RoleHash, RoleName } from '../entity/role'
+
+export const GOV_BRIDGE_ADDRESS = '0x3a7b055bf88cdc59d20d0245809c6e6b3c5819dd'
+
+export const L2_ERC20_TOKEN_GATEWAY_ADDRESS = '0x9c46560D6209743968cC24150893631A39AfDe4d'
+
+// WstETH ERC20Bridged
+export const MANTLE_WST_ETH_BRIDGED_ADDRESS = '0x458ed78EB972a369799fb278c0243b25e5242A83'
+
+export const WITHDRAWAL_INITIATED_EVENT =
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-mantle/src/utils/constants.ts:10]: I got it from here. Are sure, that are the same?
https://github.com/mantlenetworkio/lido-l2/blob/main/contracts/mantle/interfaces/IL2ERC20Bridge.sol#L11

File: l2-bridge-mantle/src/utils/constants.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  name: 'WITHDRAWALS ENABLER ROLE',
+  hash: '0x9ab8816a3dc0b3849ec1ac00483f6ec815b07eee2fd766a353311c823ad59d0d',
+}
+
+export const WITHDRAWALS_DISABLER_ROLE: Role = {
+  name: 'WITHDRAWALS DISABLER ROLE',
+  hash: '0x94a954c0bc99227eddbc0715a62a7e1056ed8784cd719c2303b685683908857c',
+}
+
+// TODO ask
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply l2-bridge-mantle/src/utils/constants.ts:51]: fix

File: l2-bridge-mantle/src/utils/constants.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (TheDZhon) [reply l2-bridge-mantle/src/utils/constants.ts:10]: Oh, it's a good catch indeed :1st_place_medal: 

Looks like the one you pasted here is correct, but Optimism and Base alerts both contain incorrect signature definition! Please update 'em.

As I see, they have the signature copy-pasted from Arbitrum, ugh.

File: l2-bridge-mantle/src/utils/constants.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+export const GOV_BRIDGE_ADDRESS = '0x3a7b055bf88cdc59d20d0245809c6e6b3c5819dd'
+
+export const L2_ERC20_TOKEN_GATEWAY_ADDRESS = '0x9c46560D6209743968cC24150893631A39AfDe4d'
+
+// WstETH ERC20Bridged
+export const MANTLE_WST_ETH_BRIDGED_ADDRESS = '0x458ed78EB972a369799fb278c0243b25e5242A83'
+
+export const WITHDRAWAL_INITIATED_EVENT =
+  'event WithdrawalInitiated( address indexed _l1Token, address indexed _l2Token, address indexed _from, address _to, uint256 _amount, bytes _data)'
```
Comment (TheDZhon) [inline l2-bridge-mantle/src/utils/constants.ts]:
```suggestion
  'event WithdrawalInitiated(address indexed _l1Token, address indexed _l2Token, address indexed _from, address _to, uint256 _amount, bytes _data)'
```

**Review by TheDZhon — CHANGES_REQUESTED** [review]
Thank you! It's the only single step is left :rocket:  (concerning the Opti/Base event signatures)

File: l2-bridge-ethereum/src/constants.ts [hunk 14]
```diff
@@ -53,10 +53,17 @@ export const OPTIMISM_L1ERC20_TOKEN_BRIDGE =
   "0x76943c0d61395d8f2edf9060e1533529cae05de6";
 export const BASE_L1ERC20_TOKEN_BRIDGE =
   "0x9de443AdC5A411E83F1878Ef24C3F52C61571e72";
-export const ZKSYNC_L1ERC20_BRIDGE =
+export const ZKSYNC_L1ERC20_TOKEN_BRIDGE =
```
Comment (TheDZhon) [inline l2-bridge-ethereum/src/constants.ts:56]: The actual contract is named as it was before this PR

```suggestion
export const ZKSYNC_L1ERC20_BRIDGE =
```

**Review by TheDZhon — CHANGES_REQUESTED** [review]
Done a review pass, and found some changes to be done

**Review by TheDZhon — APPROVED** [review]
LGTM

**Review by dgusakov — APPROVED** [review]
Approved with no detailed review

**Pull request merged**